### ISSUE TITLE
List comprehensions make far more allocations inside blocks

### ISSUE THREAD

**Issue by pranavtbhat** [JuliaLang/julia#17263]
For example:
```julia
f(x) = x

@time [f(x) for x in 1 : 10^6]
# 0.001739 seconds (2 allocations: 7.629 MB)

function gen(f::Function)
   [f(x) for x in 1 : 10^6]
end

@time gen(f);
# 0.044338 seconds (999.50 k allocations: 22.881 MB, 12.34% gc time)
```
I'm guessing the function's return type isn't being inferred?

**Comment by Keno** [comment]
Which version of julia?

**Comment by pranavtbhat** [comment]
0.4

**Comment by nalimilan** [comment]
On 0.4, function types are not inferred. This is fixed on 0.5:
```julia
julia> @time gen(f);
  0.003433 seconds (6 allocations: 7.630 MB)
```
